astroid/tree/zipper.py

```diff
@@ -0,0 +1,424 @@
+def linked_list(*values):
+    '''Builds a new linked list of tuples out of its arguments.'''
+    tail = ()
+    for value in reversed(values):
+        tail = (value, tail)
+    return tail
+
+def reverse(linked_list):
+    '''Reverses an existing linked list of tuples.'''
+    if linked_list:
+        result = collections.deque((linked_list[0],))
+        tail = linked_list[1]
+        while tail:
+            result.appendleft(tail[0])
+            tail = tail[1]
+        tail = (result.pop(), ())
+        while result:
+            tail = (result.pop(), tail)
+        return tail
+
+def iterate(linked_list):
+    '''Return an iterator over a linked list of tuples.'''
+    node = linked_list
+    while node:
+        yield node[0]
+        node = node[1]
+
+def concatenate(left, right):
+    '''Takes two existing linked lists of tuples and concatenates the
+    first one onto the second.
+
+    '''
+    if not left:
+        return right
+    elif not right:
+        return left
+    else:
+        result = [left[0]]
+        tail = left[1]
+        while tail:
+            result.append(tail[0])
+            tail = tail[1]
+        tail = (result.pop(), right)
+        while result:
+            tail = (result.pop(), tail)
+        return tail
+
+def last(linked_list):
+    '''Returns the last element of a linked list of tuples.'''
+    node = linked_list
+    while node[1]:
+        node = node[1]
+    return node[0]
+
+def initial(linked_list):
+    '''Returns a linked list of tuples containing all elements but the last.'''
+    result = [linked_list[0]]
+    tail = linked_list[1]
+    while tail:
+        result.append(tail[0])
+        tail = tail[1]
+    result.pop()
+    while result:
+        tail = (result.pop(), tail)
+    return tail
+
+
+# Attributes:
+#     left (linked list): The siblings to the left of the zipper's focus.
+#     right (linked list): The siblings to the right of the zipper's focus.
+#     parent_nodes (linked list): The ancestors of the zipper's focus
+#     parent_path (Path): The Path from the zipper that created this zipper.
+#     changed (bool): Whether this zipper has been edited or not.
+Path = collections.namedtuple('Path', 'left right parent_nodes parent_path changed')
+
+class Zipper(wrapt.ObjectProxy):
+    '''This an object-oriented version of a zipper with methods instead of
+    functions.  All the methods return a new zipper or None, and none
+    of them mutate the underlying AST nodes.  They return None when
+    the method is not valid for that zipper.  The zipper acts as a
+    proxy so the underlying node's or sequence's methods and
+    attributes are accessible through it.
+
+    Attributes:
+        __wrapped__ (base.NodeNG, collections.Sequence): The AST node or
+            sequence at the zipper's focus.
+        _self_path (Path): The Path tuple containing information about the
+            zipper's history.  This must be accessed as ._self_path.
+
+    '''
+    __slots__ = ('path')
```

Comment:
Yes.  wrapt.ObjectProxy has `__slots__` defined: https://github.com/GrahamDumpleton/wrapt/blob/develop/src/wrapt/wrappers.py#L73, and I assume the C class is similar.